### PR TITLE
extension: init tidb new extension framework

### LINKED ISSUE #38496: init the new extension framework

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: close #38496

### What is changed and how it works?

Init a new framework to allow other package to "register" extension to tidb, for example:

```
extension.Register("custom", extension.WithCustomDynPrivs([]string{"priv1", "priv2"}))
```
The extension can register it's own dynamic privileges or system variables. We'll provide more options later to make extension to extend tidb's function easily.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: extension/registry.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+		r.setup = true
+		return nil
+	}
+
+	clearBuilder := &clearFuncBuilder{}
+	defer func() {
+		if err != nil {
+			clearBuilder.Build()()
+		}
+	}()
```
Comment (xhebox) [inline extension/registry.go:89]: Isn't this duplicated with `r.close`?

Comment (lcwangchao) [reply extension/registry.go:89]: If reached line 87, `r.close` is not initialized successfully yet

**Review by xhebox — APPROVED** [review]

**Pull request merged**